compiler/AST/type.cpp

```diff
@@ -117,8 +117,25 @@ PrimitiveType::PrimitiveType(Symbol *init, bool internalType) :
 
 PrimitiveType*
 PrimitiveType::copyInner(SymbolMap* map) {
-  INT_FATAL(this, "Unexpected call to PrimitiveType::copyInner");
-  return this;
+  PrimitiveType* copy;
+
+  //
+  // If we're trying to make a copy of an internal Chapel primitive
+  // type (say 'int'), that's a sign that something is wrong.  For
+  // external primitive types, it should be OK to make such copies.
+  // This may be desired/required if the extern type declaration is
+  // local to a generic Chapel procedure for example and we're
+  // creating multiple instantiations of that procedure, each of which
+  // wants/needs its own local type symbol.  This exception may
+  // suggest that external primitive types should really be
+  // represented as their own ExternType class...
+  //
+  if (!symbol->hasFlag(FLAG_EXTERN)) {
+    INT_FATAL(this, "Unexpected call to PrimitiveType::copyInner");
+  }
+
+  copy = new PrimitiveType(NULL);
+  return copy;
 }
```

Comment:
Oh yeah, that is much better.  This is the result of saying "what do these other copyInner()s on types do, and then pattern matching.  Of course the other ones do nontrivial things with their 'copy' variable and then I didn't think about it hard enough after getting it working to note that I didn't (and didn't need to).